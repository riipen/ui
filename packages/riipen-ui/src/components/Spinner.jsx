import PropTypes from "prop-types";
import React from "react";
import _JSXStyle from "styled-jsx/style";

import withClasses from "../utils/withClasses";

import SpinnerClip from "./SpinnerClip";
import SpinnerPulse from "./SpinnerPulse";
import SpinnerScale from "./SpinnerScale";

const Spinner = ({ variant, ...other }) => {
  const spinnersMap = {
    clip: SpinnerClip,
    pulse: SpinnerPulse,
    scale: SpinnerScale
  };

  const Component = spinnersMap[variant];

  return <Component {...other} />;
};

Spinner.displayName = "Spinner";

Spinner.propTypes = {
  /**
   * The variant of spinner to display.
   */
  variant: PropTypes.oneOf(["clip", "pulse", "scale"])
};

Spinner.defaultProps = {
  variant: "scale"
};

export default withClasses(Spinner);
