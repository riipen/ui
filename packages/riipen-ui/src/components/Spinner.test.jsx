import { mount } from "enzyme";
import toJson from "enzyme-to-json";
import React from "react";

import Spinner from "./Spinner";

describe("<Spinner>", () => {
  it("renders without errors", () => {
    let error;

    try {
      mount(<Spinner />);
    } catch (theError) {
      error = theError;
    }

    expect(error).toEqual(undefined);
  });

  it("renders correct snapshot", () => {
    const wrapper = mount(<Spinner />);

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  describe("default props", () => {
    it("sets correct default props", () => {
      const defaultProps = new Spinner().type.defaultProps;

      const wrapper = mount(<Spinner />);

      expect(wrapper.find("Spinner").props().color).toEqual(defaultProps.color);
    });
  });

  describe("children prop", () => {
    it("displays given children", () => {
      const child = <div>child</div>;

      const wrapper = mount(<Spinner>{child}</Spinner>);

      expect(wrapper.containsMatchingElement(child)).toBeTruthy();
    });

    it("does not display children when loading is true", () => {
      const child = <div>child</div>;
      const loading = true;

      const wrapper = mount(<Spinner loading={loading}>{child}</Spinner>);

      expect(wrapper.containsMatchingElement(child)).toBeFalsy();
    });
  });

  describe("color prop", () => {
    it("sets color class name with a valid color", () => {
      const color = "tertiary";
      const loading = true;

      const wrapper = mount(<Spinner color={color} loading={loading} />);

      expect(
        wrapper
          .find("Spinner")
          .childAt(0)
          .childAt(0)
          .hasClass(color)
      ).toBeTruthy();
    });

    it("gives an error when given an invalid color", () => {
      const errors = jest.spyOn(console, "error").mockImplementation();
      const color = "blue";

      mount(<Spinner color={color} />);

      expect(errors).toHaveBeenCalledTimes(1);
    });
  });

  describe("loading prop", () => {
    it("renders spinner given truthy loading prop", () => {
      const loading = true;

      const wrapper = mount(<Spinner loading={loading} />);

      expect(
        wrapper
          .find("Spinner")
          .childAt(0)
          .childAt(0)
          .hasClass("loader")
      ).toBeTruthy();
    });

    it("does not render spinner given falsey loading prop", () => {
      const loading = false;

      const wrapper = mount(<Spinner loading={loading} />);

      expect(
        wrapper
          .find("Spinner")
          .childAt(0)
          .hasClass("loader")
      ).toBeFalsy();
    });
  });
});
