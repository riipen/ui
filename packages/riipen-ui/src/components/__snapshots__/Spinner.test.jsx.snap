// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Spinner> renders correct snapshot 1`] = `
<withClasses(Spinner)>
  <Spinner
    classes={
      Array [
        "riipen",
        "riipen-spinner",
      ]
    }
    variant="scale"
  >
    <withClasses(SpinnerScale)
      classes={
        Array [
          "riipen",
          "riipen-spinner",
        ]
      }
    >
      <SpinnerScale
        classes={
          Array [
            "riipen",
            "riipen-spinnerscale",
            "riipen",
            "riipen-spinner",
          ]
        }
        color="primary"
        height={40}
        margin={5}
        width={5}
      >
        <span
          className="riipen riipen-spinnerscale riipen riipen-spinner"
        >
          <span
            className="jsx-1199731182 root primary"
          />
          <JSXStyle
            dynamic={
              Array [
                0,
                40,
                5,
                5,
                "#1f71ae",
                "#f58632",
                "#3caabb",
              ]
            }
            id="4229317132"
          />
          <span
            className="jsx-429128145 root primary"
          />
          <JSXStyle
            dynamic={
              Array [
                0.1,
                40,
                5,
                5,
                "#1f71ae",
                "#f58632",
                "#3caabb",
              ]
            }
            id="4229317132"
          />
          <span
            className="jsx-824050226 root primary"
          />
          <JSXStyle
            dynamic={
              Array [
                0.2,
                40,
                5,
                5,
                "#1f71ae",
                "#f58632",
                "#3caabb",
              ]
            }
            id="4229317132"
          />
          <span
            className="jsx-2372031767 root primary"
          />
          <JSXStyle
            dynamic={
              Array [
                0.30000000000000004,
                40,
                5,
                5,
                "#1f71ae",
                "#f58632",
                "#3caabb",
              ]
            }
            id="4229317132"
          />
          <span
            className="jsx-286773236 root primary"
          />
          <JSXStyle
            dynamic={
              Array [
                0.4,
                40,
                5,
                5,
                "#1f71ae",
                "#f58632",
                "#3caabb",
              ]
            }
            id="4229317132"
          />
        </span>
      </SpinnerScale>
    </withClasses(SpinnerScale)>
  </Spinner>
</withClasses(Spinner)>
`;
